File: solutions/2023/pughmds/javascript/day06/main.js
function parseLine(line) {
    const strings = line.split(" ");
    // Step 2: Filter out non-numeric words
    const numericWords = strings.filter(word => {
        // Check if the word is numeric and can be parsed as an integer
        const parsedInt = parseInt(word);
        return !isNaN(parsedInt) && parsedInt.toString() === word;
    });
    // Step 3: Convert numeric words to integers
    return numericWords.map(word => parseInt(word));
}

function readFile(filePath) {
    const fs = require("fs");
    try {
        // Read the file synchronously
        const data = fs.readFileSync(filePath, 'utf8');

        // Split the content into an array of lines
        const lines = data.split('\n');

        // Remove empty lines
        const nonEmptyLines = lines.filter(line => line.trim() !== '');

        return nonEmptyLines;
    } catch (error) {
        console.error(`Error reading file: ${error.message}`);
        return [];
    }
}

function Part1(filePath) {
    let total = 1;
    var data = readFile(filePath);
    const times = parseLine(data[0]);
    const distances = parseLine(data[1]);
    const races = times.map((time, index) => [time, distances[index]]);

    for (const [time, distance] of races) {
        const timings = Array.from({ length: Math.min(time, distance + 1) }, (_, i) => i);
        const successes = timings.filter(timing => (time - timing) * timing > distance);

        total *= successes.length;
    }

    return total;
}

function Part2(filePath) {
    var data = readFile(filePath);
    const time_strings = data[0].replace(/\s+/g, '').match(/\d+/g);
    const distance_strings = data[1].replace(/\s+/g, '').match(/\d+/g);

    const time = time_strings.map(value => parseInt(value)).join('');
    const distance = distance_strings.map(value => parseInt(value)).join('');

    // Takes about 2.096 seconds on puzzle input
    const successes = Array.from({ length: Math.min(time, distance + 1) }, (_, timing) => timing)
        .filter(timing => timing * (time - timing) > distance);
    return successes.length;
}


var filePath = "../../inputs/day06/input.txt"
var result = Part1(filePath);
console.log(result);
var result = Part2(filePath);
console.log(result);

File: solutions/2023/pughmds/ruby/day06/main.rb
def parse_line(line)
  strings = line.split(" ")
  # Step 2: Filter out non-numeric words
  numeric_words = strings.select { |word| word.match?(/^\d+$/) }
  # Step 3: Convert numeric words to integers
  numeric_words.map(&:to_i)
end

def read_file(file_path)
  begin
    # Read the file synchronously
    data = File.read(file_path, encoding: 'utf-8')

    # Split the content into an array of lines
    lines = data.split("\n")

    # Remove empty lines
    non_empty_lines = lines.reject { |line| line.strip.empty? }

    return non_empty_lines
  rescue StandardError => e
    # kind of like the python try/except function?
    puts "Error reading file: #{e.message}"
    return []
  end
end

def part1(file_path)
  total = 1
  data = read_file(file_path)
  times = parse_line(data[0])
  distances = parse_line(data[1])
  # Pair up the times & distances
  races = times.zip(distances)

  races.each do |time, distance|
    timings = (0...[time, distance + 1].min).to_a
    successes = timings.select { |timing| (time - timing) * timing > distance }

    total *= successes.length
  end

  total
end

def part2(file_path)
  data = read_file(file_path)
  # Combine the numbers together to make a big number, removing spaces
  # Gsub swaps spaces with blanks, and scan looks for numbers in the leftover string
  time_strings = data[0].gsub(/\s+/, '').scan(/\d+/)
  distance_strings = data[1].gsub(/\s+/, '').scan(/\d+/)

  # This may not be needed as I could probably do this on the above definitions. Left here because it works (tm)
  time = time_strings.join('').to_i
  distance = distance_strings.join('').to_i

  # interesting use of the select keyword which Ruby has which works like filter. It will return values based on the
  # condition specified
  successes = (0...[time, distance + 1].min).select { |timing| timing * (time - timing) > distance }
  successes.length
end

file_path = "../../inputs/day06/input.txt"
result = part1(file_path)
puts result
result = part2(file_path)
puts result